### PR TITLE
core: pass vm.Config by reference

### PR DESCRIPTION
**Pull request by MariusVanDerWijden**
The `vm.Config` object is ~2136 bytes in size so it makes sense to pass it by reference not by value.
This greatly reduces the amount of bytes we need to allocate for every call.

Master:
```
BenchmarkInsertChain_ring1000_diskdb-16    	      79	  13206599 ns/op	 5242574 B/op	   38345 allocs/op
BenchmarkInsertChain_ring1000_memdb-16    	      75	  13402992 ns/op	 5164621 B/op	   39069 allocs/op
```
This branch:
```
BenchmarkInsertChain_ring1000_diskdb-16    	      98	  13090816 ns/op	 4709312 B/op	   38572 allocs/op
BenchmarkInsertChain_ring1000_memdb-16    	      91	  12833283 ns/op	 4700052 B/op	   39268 allocs/op
```

### CODE REVIEW

**Comment by MariusVanDerWijden** [comment]
Sorry for pinging everyone, I thought this was ready, but I have to work some more on it.

**Comment by MariusVanDerWijden** [comment]
Screenshot_2020-09-14_18-32-42

Bench03: this pr, bench04: master
It seems that this PR is slightly slower but within ~2% of master. I'm not sure if this is because of differences in machines/peers or raw performance. Since the synthetic benchmarks suggest a slight speedup I think we could still go for it

File: core/blockchain.go [hunk 1]
```diff
@@ -215,7 +215,7 @@ type BlockChain struct {
 // NewBlockChain returns a fully initialised block chain using information
 // available in the database. It initialises the default Ethereum Validator and
 // Processor.
-func NewBlockChain(db ethdb.Database, cacheConfig *CacheConfig, chainConfig *params.ChainConfig, engine consensus.Engine, vmConfig vm.Config, shouldPreserve func(block *types.Block) bool, txLookupLimit *uint64) (*BlockChain, error) {
+func NewBlockChain(db ethdb.Database, cacheConfig *CacheConfig, chainConfig *params.ChainConfig, engine consensus.Engine, vmConfig *vm.Config, shouldPreserve func(block *types.Block) bool, txLookupLimit *uint64) (*BlockChain, error) {
 	if cacheConfig == nil {
```
Comment (MariusVanDerWijden) [inline core/blockchain.go:219]: add nil check

**Comment by adamschmideg** [comment]
- [ ] Add nil check

**Comment by fjl** [comment]
Please make it possible to pass a nil `*vm.Config` to `core.NewBlockChain`. Alternatively, since we're changing the constructor already, maybe we can figure out a way to remove this parameter.

**Comment by adamschmideg** [comment]
An alternative solution is #21672

**Comment by MariusVanDerWijden** [comment]
replaced by https://github.com/ethereum/go-ethereum/pull/21672

**Pull request closed without merging**